python/pyiceberg/types.py

```diff
@@ -34,27 +34,33 @@
 import re
 from typing import (
     Any,
+    Callable,
     ClassVar,
     Dict,
     Generator,
     Literal,
     Optional,
     Tuple,
+    Type,
+    TypeVar,
 )
 
-from pydantic import Field, PrivateAttr
-from pydantic.typing import AnyCallable
+from pydantic import (
+    Field,
+    PrivateAttr,
+    RootModel,
+    model_validator,
+)
 
 from pyiceberg.typedef import IcebergBaseModel
 from pyiceberg.utils.parsing import ParseNumberFromBrackets
-from pyiceberg.utils.singleton import Singleton
 
 DECIMAL_REGEX = re.compile(r"decimal\((\d+),\s*(\d+)\)")
 FIXED = "fixed"
 FIXED_PARSER = ParseNumberFromBrackets(FIXED)
 
 
-class IcebergType(IcebergBaseModel, Singleton):
+class IcebergType(IcebergBaseModel):
```

Comment:
Is this a change we should apply anyway since not all `IcebergType` subclasses are `Singleton`?

Reply:
Currently, this is the case. The singleton will hash the arguments and if there is already an instance with the same arguments, it will be returned. I'm happy to change this to exclude:

- NestedField
- StructType
- MapType
- ListType

Also, the FixedType, DecimalType, etc could be excluded.

Reply:
(I've removed this singleton temporarily to make sure that it didn't interfere with Pydantic 2.0)